src/librustc/mir/interpret/error.rs

```diff
@@ -309,8 +311,46 @@ impl<O: fmt::Debug> fmt::Debug for PanicMessage<O> {
     }
 }
 
+#[derive(Clone, RustcEncodable, RustcDecodable, HashStable)]
+pub enum InvalidProgramMessage {
+    /// Resolution can fail if we are in a too generic context
+    TooGeneric,
+    /// Cannot compute this constant because it depends on another one
+    /// which already produced an error
+    ReferencedConstant,
+    /// Abort in case type errors are reached
+    TypeckError,
+}
+
+#[derive(Clone, RustcEncodable, RustcDecodable, HashStable)]
+pub enum UndefinedBehaviourMessage {
+}
+
+#[derive(Clone, RustcEncodable, RustcDecodable, HashStable)]
+pub enum UnsupportedMessage {
+}
+
+#[derive(Clone, RustcEncodable, RustcDecodable, HashStable)]
+pub enum ResourceExhaustionMessage {
+}
+
 #[derive(Clone, RustcEncodable, RustcDecodable, HashStable)]
 pub enum InterpError<'tcx> {
+    /// The program panicked.
+    Panic(PanicMessage<u64>),
+    /// The program caused undefined behavior.
+    UndefinedBehaviour(UndefinedBehaviourMessage),
```

Comment:
Oh also, as discussed in the issue, please rename `*Message` to `*Info`.

Reply:
Will do this.